### PR TITLE
wazuh-logcollector: add state/statistics generation.

### PR DESCRIPTION
**Pull request by jnasselle**
|Related issue|
|---|
|#6159|
## Short description

- Active state/statistics generated during `wazuh-logcollector` lifetime
- Information always accesible remotely, and optionally from a file in every agent according to a configurable time
- Historical and last interval state (optional according to a configurable time) generated periodically
- Byte, event and drop count for each active `location` and it's `targets`.

## Description

Hi team!

This PR aims to add state/statistics support for `wazuh-logcollector` during process lifetime, always accesible via remote request from the manager and also persist optionally in a file `WAZUH_DIR/var/run/wazuh-logcollector.state` (UNIX-based) or `WAZUH_DIR\wazuh-logcollector.state`. Both ways (remote request and file) are JSON formatted outputs.

One of the more important aspects of this state/statistics is the `drop` value for the different targets (`sockets`) available for a `location` and `localfile`, giving the possibility to get if the message queue is full or the connection is not successful. 

State/statistics are generated and updated every `logcollector.state_interval` seconds (internal configuration) and accesible any moment. This statistic are are divided in two sections:

- global/historical: since the startup of the process
- interval: since the last generated state info. Only if `logcollector.state_interval` is bigger than zero

If `logcollector.state_interval` is zero, state file will be disabled and remote request will still be accesible but only bringing global stats in realtime.

The state mechanism support following up to 200200 files stats, doubling the max value of `logcollector.max_files`.

### Using `getconfig` as reference for implementation
Currently, the mechanism to [Get active configuration](https://documentation.wazuh.com/4.0/user-manual/api/reference.html#operation/api.controllers.agents_controller.get_agent_config) and it's [low level implementation](https://github.com/wazuh/wazuh/blob/412981d8a0b9e5f11c6d59f3ffef08d73a989211/framework/wazuh/core/configuration.py#L752) is quite similar that the one needed by this implementation
### Data flow proposal
- Manager -> Agent : `API -> /var/ossec/queue/ossec/request -> wazuh-remoted -> wazuh-agentd ->  /var/ossec/queue/ossec/logcollector -> wazuh-logcollector`
- Manager -> Manager: `API -> /var/ossec/queue/ossec/logcollector -> wazuh-logcollector`
### Request proposal
The fact that both `wazuh-remoted` and `wazuh-agentd` need to route request along different components, the request should be the same like used in getconfig:
```
Message format: <agent-id> <module> <command> <configuration>(optional)
```
State information does not have `configuration`, so should be ommited. secure-tcp.py script will be used to exemplify a request

## Examples
[secure-tcp.py](https://github.com/wazuh/wazuh-tools/blob/master/utils/secure-tcp.py) script will be used to get state info independently from API support (#7128).

### Get agents's wazuh-logcollector `state` remotely
```
echo -n "001 logcollector getstate" | ./wazuh-tools/utils/secure-tcp.py /var/ossec/queue/ossec/request
{"error":0,"data":{"global":{"start":"Tue Jan 12 23:48:03 2021","end":"Tue Jan 12 23:48:08 2021","files":[{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/testfile","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/jira-backups.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/syslog","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/vsftpd.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"df -P","bytes":1108,"events":11},{"targets":[{"name":"agent","drops":0}],"location":"last -n 20","bytes":1518,"events":1},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/kern.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/ossec/logs/active-responses.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/audit/audit.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/dpkg.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/auth.log","bytes":0,"events":0}]},"interval":{"start":"Tue Jan 12 23:48:03 2021","end":"Tue Jan 12 23:48:08 2021","files":[{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/testfile","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/jira-backups.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/syslog","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/vsftpd.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"df -P","bytes":1108,"events":11},{"targets":[{"name":"agent","drops":0}],"location":"last -n 20","bytes":1518,"events":1},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/kern.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/ossec/logs/active-responses.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/audit/audit.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/dpkg.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/auth.log","bytes":0,"events":0}]}}}
```
### Get manager's wazuh-logcollector `state`

```
# Message format: <agent-id> <module> <command> <configuration>(optional)
echo -n "getstate" |  ./secure-tcp.py /var/ossec/queue/ossec/logcollector
{"error":0,"data":{"global":{"start":"Tue Jan 12 23:48:03 2021","end":"Tue Jan 12 23:48:08 2021","files":[{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/testfile","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/jira-backups.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/syslog","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/vsftpd.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"df -P","bytes":1108,"events":11},{"targets":[{"name":"agent","drops":0}],"location":"last -n 20","bytes":1518,"events":1},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/kern.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/ossec/logs/active-responses.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/audit/audit.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/dpkg.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/auth.log","bytes":0,"events":0}]},"interval":{"start":"Tue Jan 12 23:48:03 2021","end":"Tue Jan 12 23:48:08 2021","files":[{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/testfile","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/jira-backups.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0},{"name":"custom_socket","drops":0}],"location":"/var/log/syslog","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/vsftpd.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"df -P","bytes":1108,"events":11},{"targets":[{"name":"agent","drops":0}],"location":"last -n 20","bytes":1518,"events":1},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/kern.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/ossec/logs/active-responses.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/audit/audit.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/dpkg.log","bytes":0,"events":0},{"targets":[{"name":"agent","drops":0}],"location":"/var/log/auth.log","bytes":0,"events":0}]}}}
```
### `wazuh-logcollector.state` anatomy

```json
{
        "global":       {
                "start":        "Tue Jan 12 23:48:03 2021",
                "end":  "Wed Jan 13 01:42:22 2021",
                "files":        [{
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }, {
                                                "name": "custom_socket",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/testfile",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/jira-backups.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }, {
                                                "name": "custom_socket",
                                                "drops":        76
                                        }],
                                "location":     "/var/log/syslog",
                                "bytes":        55344,
                                "events":       76
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/vsftpd.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "wazuh-logcollector",
                                "bytes":        56,
                                "events":       1
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "df -P",
                                "bytes":        7756,
                                "events":       77
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "last -n 20",
                                "bytes":        10626,
                                "events":       7
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/kern.log",
                                "bytes":        660,
                                "events":       4
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/ossec/logs/active-responses.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/audit/audit.log",
                                "bytes":        796,
                                "events":       4
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/dpkg.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/auth.log",
                                "bytes":        0,
                                "events":       0
                        }]
        },
        "interval":     {
                "start":        "Wed Jan 13 01:42:17 2021",
                "end":  "Wed Jan 13 01:42:22 2021",
                "files":        [{
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }, {
                                                "name": "custom_socket",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/testfile",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/jira-backups.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }, {
                                                "name": "custom_socket",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/syslog",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/vsftpd.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "wazuh-logcollector",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "df -P",
                                "bytes":        1108,
                                "events":       11
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "last -n 20",
                                "bytes":        1518,
                                "events":       1
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/kern.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/ossec/logs/active-responses.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/audit/audit.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/dpkg.log",
                                "bytes":        0,
                                "events":       0
                        }, {
                                "targets":      [{
                                                "name": "agent",
                                                "drops":        0
                                        }],
                                "location":     "/var/log/auth.log",
                                "bytes":        0,
                                "events":       0
                        }]
        }
}

```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] ~MAC OS X~ (Tested on FreeBSD)
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [x] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [x] Scan-build report
  - [x] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [x] Configuration on demand reports new parameters
- [x] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

Regards,
Nico

### CODE REVIEW

**Review by juliancnn — APPROVED** [review]

**Comment by jnasselle** [comment]
## Scan-build
### Tool specs
- i686-w64-mingw32-gcc (GCC) 9.3-win32 20200320
- gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
- clang version 7.0.1-12 (tags/RELEASE_701/final)
- Ubuntu 20.04.1 LTS
### Results
[linux_agent_debug.zip](https://github.com/wazuh/wazuh/files/5818287/linux_agent_debug.zip)
[linux_agent_release.zip](https://github.com/wazuh/wazuh/files/5818288/linux_agent_release.zip)
[linux_manager_debug.zip](https://github.com/wazuh/wazuh/files/5818289/linux_manager_debug.zip)
[linux_manager_release.zip](https://github.com/wazuh/wazuh/files/5818290/linux_manager_release.zip)
[winagent_debug.zip](https://github.com/wazuh/wazuh/files/5818291/winagent_debug.zip)
[winagent_release.zip](https://github.com/wazuh/wazuh/files/5818292/winagent_release.zip)

**Comment by jnasselle** [comment]
## Unit tests
### Manager
```
./wazuh/src/unit_tests/build/logcollector/test_state
[==========] Running 29 test(s).
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_global
[       OK ] test_w_logcollector_state_init_fail_hash_create_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_interval
[       OK ] test_w_logcollector_state_init_fail_hash_create_interval
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_global
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_interval
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_interval
[ RUN      ] test_w_logcollector_state_init_ok
[       OK ] test_w_logcollector_state_init_ok
[ RUN      ] test_w_logcollector_state_get_null
[       OK ] test_w_logcollector_state_get_null
[ RUN      ] test_w_logcollector_state_get_non_null
[       OK ] test_w_logcollector_state_get_non_null
[ RUN      ] test__w_logcollector_generate_state_fail_get_node
[       OK ] test__w_logcollector_generate_state_fail_get_node
[ RUN      ] test__w_logcollector_generate_state_one_target
[       OK ] test__w_logcollector_generate_state_one_target
[ RUN      ] test__w_logcollector_generate_state_one_target_restart
[       OK ] test__w_logcollector_generate_state_one_target_restart
[ RUN      ] test__w_logcollector_state_update_file_new_data
[       OK ] test__w_logcollector_state_update_file_new_data
[ RUN      ] test__w_logcollector_state_update_file_update
[       OK ] test__w_logcollector_state_update_file_update
[ RUN      ] test__w_logcollector_state_update_file_fail_update
[       OK ] test__w_logcollector_state_update_file_fail_update
[ RUN      ] test_w_logcollector_state_update_file_null
[       OK ] test_w_logcollector_state_update_file_null
[ RUN      ] test_w_logcollector_state_update_file_ok
[       OK ] test_w_logcollector_state_update_file_ok
[ RUN      ] test__w_logcollector_state_update_target_get_file_stats_fail
[       OK ] test__w_logcollector_state_update_target_get_file_stats_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_fail
[       OK ] test__w_logcollector_state_update_target_find_target_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_ok
[       OK ] test__w_logcollector_state_update_target_find_target_ok
[ RUN      ] test__w_logcollector_state_update_target_dropped_true
[       OK ] test__w_logcollector_state_update_target_dropped_true
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Update_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Update_fail
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Add_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Add_fail
[ RUN      ] test_w_logcollector_state_update_target_null_path
[       OK ] test_w_logcollector_state_update_target_null_path
[ RUN      ] test_w_logcollector_state_update_target_null_target
[       OK ] test_w_logcollector_state_update_target_null_target
[ RUN      ] test_w_logcollector_state_update_target_ok
[       OK ] test_w_logcollector_state_update_target_ok
[ RUN      ] test_w_logcollector_generate_state_ok
[       OK ] test_w_logcollector_generate_state_ok
[ RUN      ] test_w_logcollector_state_dump_fail_open
[       OK ] test_w_logcollector_state_dump_fail_open
[ RUN      ] test_w_logcollector_state_dump_fail_write
[       OK ] test_w_logcollector_state_dump_fail_write
[ RUN      ] test_w_logcollector_state_dump_ok
[       OK ] test_w_logcollector_state_dump_ok
[ RUN      ] test_w_logcollector_state_main
[       OK ] test_w_logcollector_state_main
[==========] 29 test(s) run.
[  PASSED  ] 29 test(s).
```
```
./wazuh/src/unit_tests/build/logcollector/test_lccom
[==========] Running 2 test(s).
[ RUN      ] test_lccom_getstate_ok
[       OK ] test_lccom_getstate_ok
[ RUN      ] test_lccom_getstate_null
[       OK ] test_lccom_getstate_null
[==========] 2 test(s) run.
[  PASSED  ] 2 test(s).
```
## Linux agent
```
./wazuh/src/unit_tests/build/logcollector/test_state
[==========] Running 29 test(s).
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_global
[       OK ] test_w_logcollector_state_init_fail_hash_create_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_create_interval
[       OK ] test_w_logcollector_state_init_fail_hash_create_interval
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_global
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_global
[ RUN      ] test_w_logcollector_state_init_fail_hash_setsize_interval
[       OK ] test_w_logcollector_state_init_fail_hash_setsize_interval
[ RUN      ] test_w_logcollector_state_init_ok
[       OK ] test_w_logcollector_state_init_ok
[ RUN      ] test_w_logcollector_state_get_null
[       OK ] test_w_logcollector_state_get_null
[ RUN      ] test_w_logcollector_state_get_non_null
[       OK ] test_w_logcollector_state_get_non_null
[ RUN      ] test__w_logcollector_generate_state_fail_get_node
[       OK ] test__w_logcollector_generate_state_fail_get_node
[ RUN      ] test__w_logcollector_generate_state_one_target
[       OK ] test__w_logcollector_generate_state_one_target
[ RUN      ] test__w_logcollector_generate_state_one_target_restart
[       OK ] test__w_logcollector_generate_state_one_target_restart
[ RUN      ] test__w_logcollector_state_update_file_new_data
[       OK ] test__w_logcollector_state_update_file_new_data
[ RUN      ] test__w_logcollector_state_update_file_update
[       OK ] test__w_logcollector_state_update_file_update
[ RUN      ] test__w_logcollector_state_update_file_fail_update
[       OK ] test__w_logcollector_state_update_file_fail_update
[ RUN      ] test_w_logcollector_state_update_file_null
[       OK ] test_w_logcollector_state_update_file_null
[ RUN      ] test_w_logcollector_state_update_file_ok
[       OK ] test_w_logcollector_state_update_file_ok
[ RUN      ] test__w_logcollector_state_update_target_get_file_stats_fail
[       OK ] test__w_logcollector_state_update_target_get_file_stats_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_fail
[       OK ] test__w_logcollector_state_update_target_find_target_fail
[ RUN      ] test__w_logcollector_state_update_target_find_target_ok
[       OK ] test__w_logcollector_state_update_target_find_target_ok
[ RUN      ] test__w_logcollector_state_update_target_dropped_true
[       OK ] test__w_logcollector_state_update_target_dropped_true
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Update_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Update_fail
[ RUN      ] test__w_logcollector_state_update_target_OSHash_Add_fail
[       OK ] test__w_logcollector_state_update_target_OSHash_Add_fail
[ RUN      ] test_w_logcollector_state_update_target_null_path
[       OK ] test_w_logcollector_state_update_target_null_path
[ RUN      ] test_w_logcollector_state_update_target_null_target
[       OK ] test_w_logcollector_state_update_target_null_target
[ RUN      ] test_w_logcollector_state_update_target_ok
[       OK ] test_w_logcollector_state_update_target_ok
[ RUN      ] test_w_logcollector_generate_state_ok
[       OK ] test_w_logcollector_generate_state_ok
[ RUN      ] test_w_logcollector_state_dump_fail_open
[       OK ] test_w_logcollector_state_dump_fail_open
[ RUN      ] test_w_logcollector_state_dump_fail_write
[       OK ] test_w_logcollector_state_dump_fail_write
[ RUN      ] test_w_logcollector_state_dump_ok
[       OK ] test_w_logcollector_state_dump_ok
[ RUN      ] test_w_logcollector_state_main
[       OK ] test_w_logcollector_state_main
[==========] 29 test(s) run.
[  PASSED  ] 29 test(s).
```
```
./wazuh/src/unit_tests/build/logcollector/test_lccom
[==========] Running 2 test(s).
[ RUN      ] test_lccom_getstate_ok
[       OK ] test_lccom_getstate_ok
[ RUN      ] test_lccom_getstate_null
[       OK ] test_lccom_getstate_null
[==========] 2 test(s) run.
[  PASSED  ] 2 test(s).
```

**Review by juliancnn — APPROVED** [review]

**Review by Lopuiz — APPROVED** [review]

**Review by JcabreraC — APPROVED** [review]

**Review by juliancnn — APPROVED** [review]

**Review by nmkoremblum — APPROVED** [review]
LGTM :+1:

**Review by juliancnn — APPROVED** [review]

File: src/logcollector/lccom.c [hunk 1]
```diff
@@ -33,13 +34,32 @@ size_t lccom_dispatch(char * command, char ** output){
...
+        return lccom_getstate(output);
     } else {
         mdebug1("LCCOM Unrecognized command '%s'.", rcv_comm);
         os_strdup("err Unrecognized command", *output);
         return strlen(*output);
     }
 }
 
+size_t lccom_getstate(char ** output) {
+    cJSON * state_json = NULL;
```
Comment (nmkoremblum) [inline src/logcollector/lccom.c:47]: I would assign **_w_logcollector_state_get_**'s output here and would add the modifier "**const**" if this variable won't be modified

File: src/logcollector/lccom.c [hunk 2]
```diff
@@ -33,13 +34,32 @@ size_t lccom_dispatch(char * command, char ** output){
...
+    cJSON * w_packet = cJSON_CreateObject();
+    if (state_json = w_logcollector_state_get(), state_json == NULL) {
+        cJSON_AddNumberToObject(w_packet, "error", 1);
```
Comment (nmkoremblum) [inline src/logcollector/lccom.c:50]: I would replace **errors ID** with "self documented" **_defines_** for better readability

File: src/logcollector/state.c [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    char * lc_state_str = cJSON_Print(lc_state_json);
+    cJSON_Delete(lc_state_json);
+
+    // Add trailing newline
+    const size_t len = strlen(lc_state_str);
+    os_realloc(lc_state_str, len + 2, lc_state_str);
+    lc_state_str[len] = '\n';
+    lc_state_str[len + 1] = '\0';
+
+    FILE * lc_state_file = NULL;
```
Comment (nmkoremblum) [inline src/logcollector/state.c:109]: I would assign the **_fopen_**'s output here and would add the "**_const_**" modifier

**Review by nmkoremblum — APPROVED** [review]
LGTM :)

**Review by JcabreraC — APPROVED** [review]

**Pull request merged**